my_2d_robot/src/calibration_2d.cpp

```diff
@@ -0,0 +1,130 @@
+    residual[1] = T(link1_ * sin(offset_joint1_ - offset1[1]) +
+                    link2_ * sin(offset_joint1_ - offset1[1] + offset_joint2_ - offset2[1])) -
+                  true_y_;
+
+    return true;
+  }
+
+private:
+  // Accurate position of end effector.
+  const double true_x_;
+  const double true_y_;
+  // Offset joint angles.
+  double offset_joint1_;
+  double offset_joint2_;
+  // Link lengths.
+  double link1_;
+  double link2_;
+};
+
+// Function that reads yaml file containing joint angles and end effector position data, stores the data in vectors, and
+// uses ceres solver to calculate the joint angle offsets.
+void readYaml()
+{
+  // Filepath of file to use (specified in launch file).
+  std::string filepath;
+  // Number of data points written to yaml file by subscriber node.
+  int data_num_max_;
+  ros::NodeHandle n;
+  n.getParam("data_point_count", data_num_max_);
+  n.param<std::string>("file_path", filepath, "did not work this time");
+
+  // Vector to store the offset joint angles.
+  std::vector<double> angle_yaml;
+  // Vector to store end effector positions.
+  std::vector<double> position_yaml;
+
+  YAML::Node data_yaml = YAML::LoadFile(filepath);
+
+  const YAML::Node& data = data_yaml["data"];
+
+  for (YAML::const_iterator it = data.begin(); it != data.end(); ++it)
+  {
+    const YAML::Node& reading = *it;
+
+    // Storing in vectors using for loop.
+    angle_yaml.push_back(reading["joint_angles"][0].as<double>());
+    angle_yaml.push_back(reading["joint_angles"][1].as<double>());
+    position_yaml.push_back(reading["end_effector_position"][0].as<double>());
+    position_yaml.push_back(reading["end_effector_position"][1].as<double>());
+  }
+
+  // Printing the offset joint angles and actual end effector positions.
+  for (int i = 0; i < data_num_max_; i++)
+  {
+    std::cout << "(" << angle_yaml[2 * i] << "," << angle_yaml[2 * i + 1] << ")" << std::endl;
+    std::cout << "(" << position_yaml[2 * i] << "," << position_yaml[2 * i + 1] << ")" << std::endl;
+  }
+
+  // Initial values of offsets for ceres solver.
+  double offset1 = 0.0;
+  double offset2 = 0.0;
+
+  ceres::Problem problem;
+  for (int i = 0; i < data_num_max_; i++)
+  {
+    problem.AddResidualBlock(
+        new AutoDiffCostFunction<OffsetCalibration, 1, 1, 1>(new OffsetCalibration(
+            angle_yaml[2 * i], angle_yaml[2 * i + 1], position_yaml[2 * i], position_yaml[2 * i + 1])),
+        NULL, &offset1, &offset2);
+  }
+
+  ceres::Solver::Options options;
+  options.linear_solver_type = ceres::DENSE_QR;
+  options.minimizer_progress_to_stdout = true;
+
+  ceres::Solver::Summary summary;
+  Solve(options, &problem, &summary);
+  std::cout << summary.BriefReport() << "\n";
+  std::cout << "Offset 1 : " << offset1 << "\n";
+  std::cout << "Offset 2 : " << offset2 << "\n";
+}
+
+int main(int argc, char** argv)
+{
+  google::InitGoogleLogging(argv[0]);
+
+  ros::init(argc, argv, "robo_2d_calibrator");
+
+  readYaml();
```

Comment:
From a design perspective, I get that you are calling the ceres functions in readYaml. But we should think more carefully about the meaning behind our function names.

`readYaml` here does alot more than just reading a YAML file. So maybe we should change this name to `calibrate`

`main` should do all the variable parsing from the parameter server. Then pass these vars as input to calibrate.

`calibrate` should just accept variables  like:
- data_num_max_
- filepath
and do all the data reading from the file and the actual calibration.